### DISCUSSION TITLE
Understanding queues and nodes

### DISCUSSION THREAD

**Discussion by wlandau** [r-lib/mirai#22] [Q&A] [answered]
As discussed in #18, I would like to use https://github.com/shikokuchuo/mirai#example-1-connecting-to-remote-servers--remote-server-queues to distribute tasks across jobs on a cluster (and eventually, across AWS Batch jobs on the same local network). Each cluster or AWS Batch job will run one R process that calls `mirai::server()` which connects back to the client, and I assume each server process will accept one task at a time. Tasks will vary in runtime, and there will be more tasks than servers/nodes, so it sounds like an active queue will be useful.

I have learned a lot from the README, the documentation of `daemons()` and `server()`, my experiments with the code, and your helpful comments in #18. However, I still do not understand how queueing or nodes work in https://github.com/shikokuchuo/mirai#example-1-connecting-to-remote-servers--remote-server-queues. I would appreciate it if we could revisit the topic at a beginner level. My code examples below use commit 812e93f9cdc78670849ca48b89dffe8f58e8fdc8. I apologize if my questions repeat things I already asked in #19.

## What I think I understand so far

Aside from autoscaling and crash detection, here is how I see my use case playing out. The following example has 2 workers and only a few tasks, but I envision scaling up to hundreds of each.

First, I open a TCP socket on the client:

```r
# from xx.xx.xx.175
daemons("tcp://xx.xx.xx.175:5555")
```

and then on a different computer on the network I run a server process

```r
# from xx.xx.xx.138
server("tcp://xx.xx.xx.175:5555")
```

and on a third computer I run a third server process

```r
# from xx.xx.xx.176
server("tcp://xx.xx.xx.175:5555")
```

Then the client can send jobs for the two servers to run. Below, I experiment with a task load of 6 jobs for 2 servers.

```r
# from xx.xx.xx.175
tasks <- list(
  mirai({Sys.sleep(5); paste0("task1_server", gsub("^.*\\.", "", getip::getip()))}),
  mirai({Sys.sleep(3); paste0("task2_server", gsub("^.*\\.", "", getip::getip()))}),
  mirai({Sys.sleep(4); paste0("task3_server", gsub("^.*\\.", "", getip::getip()))}),
  mirai({Sys.sleep(2); paste0("task4_server", gsub("^.*\\.", "", getip::getip()))}),
  mirai({Sys.sleep(8); paste0("task5_server", gsub("^.*\\.", "", getip::getip()))}),
  mirai({Sys.sleep(7); paste0("task6_server", gsub("^.*\\.", "", getip::getip()))})
)
for (i in seq_len(60)) {
  Sys.sleep(1)
  seconds <- paste0(i, "s")
  print(paste(c(seconds, purrr::map_chr(tasks, ~.x$data)), collapse = " "))
}
```

The output I see from the loop on the client is:

```r
[1] "1s NA NA NA NA NA NA"
[1] "2s NA NA NA NA NA NA"
[1] "3s NA task2_server138 NA NA NA NA"
[1] "4s NA task2_server138 NA NA NA NA"
[1] "5s task1_server176 task2_server138 NA task4_server138 NA NA"
[1] "6s task1_server176 task2_server138 NA task4_server138 NA NA"
[1] "7s task1_server176 task2_server138 NA task4_server138 NA NA"
[1] "8s task1_server176 task2_server138 NA task4_server138 NA NA"
[1] "9s task1_server176 task2_server138 task3_server176 task4_server138 NA NA"
[1] "10s task1_server176 task2_server138 task3_server176 task4_server138 NA NA"
[1] "11s task1_server176 task2_server138 task3_server176 task4_server138 NA NA"
[1] "12s task1_server176 task2_server138 task3_server176 task4_server138 NA task6_server138"
[1] "13s task1_server176 task2_server138 task3_server176 task4_server138 NA task6_server138"
[1] "14s task1_server176 task2_server138 task3_server176 task4_server138 NA task6_server138"
[1] "15s task1_server176 task2_server138 task3_server176 task4_server138 NA task6_server138"
[1] "16s task1_server176 task2_server138 task3_server176 task4_server138 NA task6_server138"
[1] "17s task1_server176 task2_server138 task3_server176 task4_server138 task5_server176 task6_server138"
```

From this output, we can draw a picture of which tasks are running at which times. In the plot below, each row is a task. Time advances from left to right, and each task is shaded according to the server running it at a given time point.

![221080450-10632625-d998-4463-8fe2-0c4351f2bfc2](https://user-images.githubusercontent.com/1580860/221449163-b26b3aea-4918-4231-a08d-d05c7c6b3fe5.png)

I really like the way `mirai` scheduled these tasks. Even though there were more tasks than servers, and even though the tasks varied in length, each worker was as busy as possible, and tasks on the queue began as soon as possible. Unless there is something special about auto-scaling (e.g. #19 and upscaling) or the integer-length runtimes I coded, the scheduling in this part of `mirai` is already perfect. In my situation, I find it hard to see why I would need an active queue or the "nodes" functionality.  (Admittedly, I do not have grasp of the `mirai`-specific versions of these concepts.) Am I missing something relevant to my scenario?

# Nodes and queues?

In https://github.com/shikokuchuo/mirai#example-1-connecting-to-remote-servers--remote-server-queues, you explain that `server()` accepts a `nodes` argument. That section and https://github.com/shikokuchuo/mirai#example-2-connecting-to-remote-servers-through-a-local-server-queue describe active queues. So my questions are:

1. What exactly does the `nodes` argument do?
2. What exactly is an active queue in `mirai`, and how do it work? (I.e. where does it run and what does it orchestrate?)
3. Given the positive outcome from my code example above, do I even need (1) or (2)?

I am sorry if I am missing something obvious from the code usage examples, documentation, or #18, but would you be willing to explain it to me at a more basic level what these features are and the specific scenarios that motivate them?

**Comment by shikokuchuo** [comment, marked as answer]
Let me answer (3) first – if you don’t need an active queue, don’t use it. The underlying NNG logic and implementation is very robust and I have recommended throughout the documentation that if this is suitable then it should be used.

The problem we have is that when we send the tasks we have no way of knowing the task length a priori. In your example, the tasks are still roughly the same length so the solution is more or less acceptable.

Let me give you an extreme counter-example: odd number tasks length 1, even number tasks length 10. As NNG round-robins*, the odd number tasks are all sent to server 1, and the even ones to server 2. Server 1 will be idle after 3 seconds, However the total time taken will be 30 seconds.

This is the reason for an active queue. 

*NNG is not so dumb it also responds to back-pressure from the socket, but only in the case where messages are sent faster than socket buffers can be cleared. Buffers are tuneable at the NNG level but there are also system level TCP socket buffers - in short this is not something we can control reliably.

**Reply by wlandau** [reply]
> Let me give you an extreme counter-example: odd number tasks length 1, even number tasks length 10. As NNG round-robins*, the odd number tasks are all sent to server 1, and the even ones to server 2. Server 1 will be idle after 3 seconds, However the total time taken will be 30 seconds.

Ah, I see now. Thanks for explaining. It sounds like an active queue would be useful after all, and you have given me a testable example where it makes a difference. 

> This is the reason for an active queue.

Yes, I agree now. Thank you for patiently explaining it to me.

**Comment by shikokuchuo** [comment]
(1) Specifying `nodes` means using an active server queue. It will launch on the local machine a background process running this branch of the `server()` code: https://github.com/shikokuchuo/mirai/blob/main/R/mirai.R#L84

**Reply by wlandau** [reply]
In that case, what does `nodes` mean in terms of how the active server queue works? Say I have a pipeline that auto-scales between 4 and 8 servers, depending on the workload. What value should I set `nodes` to be in the call to `daemons()`?

**Reply by shikokuchuo** [reply]
In a new commit 51adb17 v0.7.2.9007 - this is now possible. You can set `nodes = 8` and auto-scale your servers however you want.

**Comment by shikokuchuo** [comment]
(2) What the active queue does is act as a relay or switch - it sits in the middle and forwards tasks between the client and an end server.

The logic it has is to only forward tasks to servers that are free (idle). This is the key difference! As we cannot know a priori the task length, we should not allocate them all at once - they need to be queued. Now we just need to poll for servers becoming free and send tasks to the server if there are remaining tasks in the queue. The polling is why we run this in a background process.

So if you now call:
```
daemons("tcp://:5555", n = 2)
```
(I find it easier to specify all interfaces on the client, also partial matching works so 'n' is as good as 'nodes')

and 
```
# from xx.xx.xx.138
server("tcp://xx.xx.xx.175:5555")
```

```
# from xx.xx.xx.176
server("tcp://xx.xx.xx.175:5556")
```
[]* Note the increment of the port number

You will see now that waiting tasks will be allocated to a server as soon as it becomes available.

**Reply by wlandau** [reply]
Thanks for confirming. Is it possible to have an active server queue, distributing tasks to available servers, while only occupying a single port number? Does the active server queue always require one port per server, as opposed to all servers sharing the same port? As discussed before, my situation would benefit from an active queue, but I do not think I can afford to occupy so many ports. One port per server group in #23 would be fine, but I plan to launch hundreds of servers, and I worry about exhausting the port range, especially on shared machines.

**Reply by wlandau** [reply]
>(2) What the active queue does is act as a relay or switch - it sits in the middle and forwards tasks between the client and an end server.
> 
> The logic it has is to only forward tasks to servers that are free (idle). This is the key difference! As we cannot know a priori the task length, we should not allocate them all at once - they need to be queued. Now we just need to poll for servers becoming free and send tasks to the server if there are remaining tasks in the queue. The polling is why we run this in a background process.

I am imagining that this background process runs on the same machine as the client, and it periodically checks the TCP socket to see if servers are free to accept jobs. This seems like something the client machine does.  I am having trouble squaring that fact with this part of the README:

> Alternatively, supply ‘nodes’ to launch an active server queue directing a cluster with the specified number of nodes: mirai::server("tcp://192.168.0.2:5555",nodes = 8)

If the active queue is a background process on the client, and the `nodes` argument activates an active queue, why does `server()` need a `nodes` argument? How is that different from calling `daemons()` with `nodes`, and what exactly is each node?

**Reply by wlandau** [reply]
Would this kind of active queue work?

1. Poll the number of active connections to the TCP socket.
2. Poll the number of busy servers among those connections.
3. If (1) is strictly greater than (2), submit more tasks (to the idle servers).

Is this kind of active queue possible while allowing all servers to share the same port?

**Reply by wlandau** [reply]
I feel like I am definitely missing something about how `server(nodes = 8)` works. Because that version of the active queue does not need one port per server, right? I am stil confused because I do not know what it means to set the number of nodes inside one of those very same nodes.

**Reply by shikokuchuo** [reply]
> Would this kind of active queue work?
> 
>     1. Poll the number of active connections to the TCP socket.
> 
>     2. Poll the number of busy servers among those connections.
> 
>     3. If (1) is strictly greater than (2), submit more tasks (to the idle servers).
> 

This is implemented in 51adb17 v0.7.2.9007. You can now set `nodes` to be the max number of nodes. The queue will then adjust to the number of actual connections.

**Reply by shikokuchuo** [reply]
> I feel like I am definitely missing something about how `server(nodes = 8)` works. Because that version of the active queue does not need one port per server, right? I am stil confused because I do not know what it means to set the number of nodes inside one of those very same nodes.

In short `daemons(URL, nodes = 8)` does not simply pass the URL argument through. I have overloaded the argument... it also provides the URL for connecting itself with the client.

`server(URL, nodes = 8)` will launch an active queue along with the nodes on the same machine (server queue / server nodes setup)

It may help to think of `daemons()` as purely a client function. `server()` as purely a server function. Each function can only do things on its own machine.

**Reply by shikokuchuo** [reply]
> Thanks for confirming. Is it possible to have an active server queue, distributing tasks to available servers, while only occupying a single port number? Does the active server queue always require one port per server, as opposed to all servers sharing the same port? As discussed before, my situation would benefit from an active queue, but I do not think I can afford to occupy so many ports. One port per server group in #23 would be fine, but I plan to launch hundreds of servers, and I worry about exhausting the port range, especially on shared machines.

The port range is for your own client. Each server will only need one outbound port. Even with hundreds of connections, I shouldn't think it would be an issue to use a range high in the 65,536 available ports.

Using any other solution will be sub-optimal - more complicated and you will have to sacrifice guaranteed delivery.

**Reply by wlandau** [reply]
Yes, you're right, the client does only take up a single port, and same with each server. More of an option than I originally though. Still not sure how I feel about the limited port range though, and when it comes to auto-scaling with e.g. `server(idletime = 60000)`, I would need to figure out which ports to use for scaling back up the workers.

On reflection, the client loop in `targets` is already prepared to deal with a passive queue (current approach with `clustermq` via `tar_make_clustermq()`). So I might actually go with that after all, I will need to think about it some more. On the one hand, it would be nice to have an active queue for non-`targets` use cases such as highly interactive Shiny apps. On the other, the user still needs to periodically check the queue anyway, whether passive or active.

**Reply by shikokuchuo** [reply]
> Yes, you're right, the client does only take up a single port, and same with each server. More of an option than I originally though. Still not sure how I feel about the limited port range though, and when it comes to auto-scaling with e.g. `server(idletime = 60000)`, I would need to figure out which ports to use for scaling back up the workers.

If using an active queue, the client listeners will be bound to the same set of ports (on the client) regardless of what your servers are doing. So you should be able to freely scale up and down servers connecting into the *same* ports.

**Reply by shikokuchuo** [reply]
Just in case the terminology is causing confusion - the client *listens* on ports on the client (specified by you, or assigned by NNG if specifying port 0). The server *dials* into one of those ports on the client (you don't care using what outbound port, that is outside of your control). The URL is always the IP address and port of the client, not the server (you never use it's address).

It is perhaps more natural to think server/listener - client/dialer - but that is precisely the wrong way around.

**Reply by HenrikBengtsson** [reply]
Coming in late to this discussion:

For my clarification, is there a one-to-one relationship to of the rows in `daemons()$daemons` and the servers?  That is, is there one, and only one, server running for each row in `daemons()$daemons` with `online = 1`?  In other words, is it correct that `online` is a "boolean" that can only take values `0` or `1`?  What does "online status" really mean here?

In `?mirai::daemons` is says that "instance number (increments each time a server connects to the URL)".  Can you give a simple example showing how you can end up with `instance > 1`?  It sounds like you can have two or more servers connected to the same daemon, but, if so, then it's clear I don't understand how daemon-server mapping works.

If my understanding of the above is correct, would it make sense to turn `daemons()$daemons` into a data frame instead?  Then `online` could be a logical taking `FALSE` and `TRUE`, making it a bit more clear what it represents. The URI could be a character column, instead of a row name. I assume, columns `instance`, `assigned`, and `complete` would all be integers (=counts).

**Reply by wlandau** [reply]
From my work on `crew`, I think I can help with some of this. @shikokuchuo, please correct me if I am wrong.

> For my clarification, is there a one-to-one relationship to of the rows in daemons()$daemons and the servers? That is, is there one, and only one, server running for each row in daemons()$daemons with online = 1? In other words, is it correct that online is a "boolean" that can only take values 0 or 1? What does "online status" really mean here?

 Each row in `daemons()$daemons` corresponds to a websocket or full TCP socket where the client or dispatcher is listening. One or more servers can connect ("dial in") to any of these websockets to receive tasks from the client. If `online` is 0, then no server is dialed in to the socket. I do not remember if `online == 1` means exactly 1 server or *at least* one server. In the former case, I would expect `online` to be able to exceed 1.

> In ?mirai::daemons is says that "instance number (increments each time a server connects to the URL)". Can you give a simple example showing how you can end up with instance > 1?

`crew` relies on this capability for auto-scaling and semi-persistent workers. When the task load is high enough, `crew` launches a server that dials into a websocket where the `mirai` client is listening. If `server()` was called with a sufficiently low `idletime`, the server will exit on its own if it idles for long enough. Later on, if the workload surges again, `crew` will launch a different server to take its place. (Although for reasons that get a bit more complicated, `crew` needs to call `mirai::saisei()` on each relaunch in order to rotate the suffix of the websocket.)

> If my understanding of the above is correct, would it make sense to turn daemons()$daemons into a data frame instead? Then online could be a logical taking FALSE and TRUE, making it a bit more clear what it represents. The URI could be a character column, instead of a row name. I assume, columns instance, assigned, and complete would all be integers (=counts).

I think a data frame is a good fit conceptually. I wonder though, would that impact efficiency? Data frame operations tend to pop up a lot in my `proffer`/`pprof` flame graphs. I think usage is still clear whether it is a matrix or data frame, and I would really hope that a batch of 1000 calls to `daemons()$daemons` would not slow down.

**Reply by HenrikBengtsson** [reply]
Thanks @wlandau, this is really helpful. I'll follow up with small questions to rule out that I'm not misunderstanding things. Based on this, is it correct that:

```r
sum(daemons()$daemons[, "online"])
```

gives me the total number of parallel workers (=servers) currently running?  And, 


```r
sum(daemons()$daemons[, "assigned"])
```

gives the number of workers currently busy?


> One or more servers can connect ("dial in") to any of these websockets to receive tasks from the client.

I tried the following before asking my questions:

```sh
> mirai::daemons(1L, url = "ws://:0")
> uri <- rownames(mirai::daemons()$daemons)
> url
[1] "ws://:44591"
attr(,"dispatcher_pid")
[1] 575731
```

When I then launch a server in the background, e.g.

```sh
Rscript -e 'mirai::server("ws://localhost:44591")'
```

I can see the server connects;

```r
> mirai::daemons()$daemons
            online instance assigned complete
ws://:44591      1        1        0        0
```

But, if I launch a second server;

```sh
Rscript -e 'mirai::server("ws://localhost:44591")'
```

I get:

```r
> mirai::daemons()$daemons
            online instance assigned complete
ws://:44591      0        1        0        0
```

Why does `online` become `0` here?

Launching a third, gives:

```r
> mirai::daemons()$daemons
            online instance assigned complete
ws://:44591      1        2        0        0
```

Launching a fourth, gives:

```r
> mirai::daemons()$daemons
            online instance assigned complete
ws://:44591      0        2        0        0
```

These results confuse me.


> I would really hope that a batch of 1000 calls to `daemons()$daemons` would not slow down.

I'm curious, what makes you having to poll that information that much?

**Reply by shikokuchuo** [reply]
Thanks for the help @wlandau. To supplement:

1. "online" means there is an active pipe connection established (the server is fully ready to receive tasks). This is a boolean.
2. Precisely, this was implemented to enable ease of auto-scaling in `crew`. To see 'instance' in action:
```r
library(mirai)
daemons(1, maxtasks = 1L)
daemons()
m <- mirai(1)
daemons()
launch_server(row.names(daemons()$daemons)[[1L]], maxtasks = 1L)
daemons()
m <- mirai(1)
daemons()
```
3. I generally prefer using a matrix where the data allows. Although it is currently what prevents the 'online' status from being a logical value as opposed to an integer. Using a data.frame would make retrieving the URLs easier for cases such as the example in 2. In terms of construction, I can make a data.frame almost as efficiently as a matrix. In terms of extracting values, I believe it is simply a case of always using `.subset2()` to avoid method dispatch.

**Reply by shikokuchuo** [reply]
> 
> These results confuse me.
> 

There is meant to be a one to one relationship between a URL and server instance when using dispatcher. Connecting more than one server to a unique URI leads to unsupported behaviour. You will notice that tasks do not get done if the 'online' status is zero.

Just to add, I think perhaps @wlandau was thinking of daemons without dispatcher. In which case, many servers can dial into the same address. But just to be clear, when using dispatcher i.e. any time that matrix appears, it should be either one or zero.

**Reply by shikokuchuo** [reply]
> Thanks @wlandau, this is really helpful. I'll follow up with small questions to rule out that I'm not misunderstanding things. Based on this, is it correct that:
> 
> ```r
> sum(daemons()$daemons[, "online"])
> ```
> 
> gives me the total number of parallel workers (=servers) currently running? And,

Correct

> 
> ```r
> sum(daemons()$daemons[, "assigned"])
> ```
> 
> gives the number of workers currently busy?
> 
Not quite, that would be
```r
sum(daemons()$daemons[, "assigned"]) - sum(daemons()$daemons[, "complete"])
```

**Reply by shikokuchuo** [reply]
@wlandau this is what I mean by using `.subset2()` in my comment above. As a data.frame is simply a classed list, you can simply treat it as such, in which case it is not inherently slow. It would make your code slightly less elegant though :)

``` r
library(microbenchmark)
df <- data.frame(a = 1, b = 2)
df
#>   a b
#> 1 1 2
microbenchmark(df[["a"]], .subset2(df, "a"), times = 10000)
#> Unit: nanoseconds
#>               expr  min   lq      mean median   uq     max neval
#>          df[["a"]] 2662 2777 3777.1149 2983.5 4739 1182427 10000
#>  .subset2(df, "a")   80  104  132.1207  113.0  125    6625 10000
```
<sup>Created on 2023-05-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

The other issue would be to make everything compatible while `crew` switches over, if you were happy to do so.

**Reply by HenrikBengtsson** [reply]
Thank you.

> 1. "online" means there is an active pipe connection established (the server is fully ready to receive tasks). This is a boolean.
> ...
> There is meant to be a one to one relationship between a URL and server instance when using dispatcher. Connecting more than one server to a unique URI leads to unsupported behaviour. You will notice that tasks do not get done if the 'online' status is zero.

Does this mean that you can have at most `nrow(daemons()$daemons)` servers connection back ("signing in")?

---

However, I'm still confused. Trying your example, I get:

```r
> library(mirai)
> daemons(1, maxtasks = 1L)
[1] 1

> daemons()
$connections
[1] 1

$daemons
                                                    online instance assigned
abstract://682cee060c4fab023a5c9b1248b479dc621afa17      1        1        0
                                                    complete
abstract://682cee060c4fab023a5c9b1248b479dc621afa17        0
```

From the above I conclude there is one "server" connected (`instance = 1`) and there is a "pipe" ready to it (`online = 1`).

Continuing:

```r
> m <- mirai(1)
> daemons()
$connections
[1] 1

$daemons
                                                    online instance assigned
abstract://682cee060c4fab023a5c9b1248b479dc621afa17      0        1        1
                                                    complete
abstract://682cee060c4fab023a5c9b1248b479dc621afa17        1
```

From this, plus the previous state, I conclude that the task was assigned (`assigned = 1`) and completed (`complete = 1`). [BTW, should it be called `completed` instead of `complete` in parallel with `assigned`?]

From `instance = 1`, I conclude the server is still connected, but I don't understand why `online = 0` and not `online = 1`?  What changed?  Because, then `sum(daemons()$daemons[, "online"])` says there are zero workers available.

Collecting the results makes no difference;

```r
> m$data
[1] 1
> daemons()
$connections
[1] 1

$daemons
                                                    online instance assigned
abstract://38455164d96b891aea32836a026321e3c696ff87      0        1        1
                                                    complete
abstract://38455164d96b891aea32836a026321e3c696ff87        1
```

Continuing with your example, I call and get:


```r
> launch_server(row.names(daemons()$daemons)[[1L]], maxtasks = 1L)
> daemons()
$connections
[1] 1

$daemons
                                                    online instance assigned
abstract://682cee060c4fab023a5c9b1248b479dc621afa17      1        2        0
                                                    complete
abstract://682cee060c4fab023a5c9b1248b479dc621afa17        0
```

So, I'm still confused, because above you said: "Connecting more than one server to a unique URI leads to unsupported behaviour.".  Are there one or two servers running/connected at this point?  From `instance = 2`, I'd say two, but that contradicts it's only possible to have at most one.

Continuing on:

```r
> m <- mirai(2)
> daemons()
$connections
[1] 1

$daemons
                                                    online instance assigned
abstract://682cee060c4fab023a5c9b1248b479dc621afa17      0        2        1
                                                    complete
abstract://682cee060c4fab023a5c9b1248b479dc621afa17        1
```

Hmm... `online` went back to `0` again.

Just to make sure, I got the correct results back:

```r
> m$data
[1] 2
```

That's good, but if I try to launch yet another task, it does not get resolved;

```r
> m <- mirai(3)
> m$data
'unresolved' logi NA
```

It stays like this forever.  At this point I see:

```r
> daemons()
$connections
[1] 1

$daemons
                                                    online instance assigned
abstract://9667cb6e2930a6392cdd49f5b4e885104f9cfaf4      0        2        1
                                                    complete
abstract://9667cb6e2930a6392cdd49f5b4e885104f9cfaf4        1
```

I'm using mirai 0.8.3 and nanonext 0.8.2 here.

**Reply by shikokuchuo** [reply]
Yes, what you are experiencing is intended. The design was primarily to facilitate the scaling algorithms in `crew`. 

When a server disconnects, online goes from 1 to 0. The instance and task stats remain the same.
When a new server connects at the same URL, online goes from 0 to 1, instance increases by 1, and the task stats are reset to 0.

Perhaps it is not as obvious to a new user - I am happy to add to the documentation at least to make it clearer.


> Does this mean that you can have at most nrow(daemons()$daemons) servers connection back ("signing in")?

Yes, the number of daemons 'n' can be thought of as the maximum servers that can be connected at any one time. `crew` is able to scale up or down from 0 to 'n'.

> So, I'm still confused, because above you said: "Connecting more than one server to a unique URI leads to unsupported behaviour.". Are there one or two servers running/connected at this point? From instance = 2, I'd say two, but that contradicts it's only possible to have at most one.

At the point you mention, one server is connected. It is the 2nd instance of the server at that URL. That is how 'online' and 'instance' are meant to be read.

> It stays like this forever. At this point I see:

At this point, the server has tasked out and disconnected again, hence 'online' is 0, and your mirai remain unresolved.

I hope the above helps.

**Reply by HenrikBengtsson** [reply]
Thank you.

> Perhaps it is not as obvious to a new user - I am happy to add to the documentation at least to make it clearer.

Yes, I think that would be helpful.  Right now I'm not sure if I'm overthinking it, but after trying to understand what `daemons()$daemons` represents, I tried to reverse engineer it from different trial-and-error examples, and I think I ended up in a rabbit hole doing so.

Regarding `online`: So, if I ever see it go from `1` to `0`, does that indicate something went wrong, e.g. the server terminated, or the network connection went down?  Because as soon as I've got a server connected, and I don't shut it down intentionally, or it times out, it should stay at `online = 1`, correct?

Is it correct that:

```r
online <- (daemons()$daemons[, "online"] == 1L)
```

is relevant for knowing the _current_ state of the "cluster", whereas:

```r
stats <- mirai::daemons()$daemons[, c("instance", "assigned", "complete")]
```

is _historical_ data for what happened in the past (and irrelevant for task/worker orchestration)?

If you want to scale up and down the servers, is it correct that it's only the `online` flag you should look at, e.g. only launch servers for daemons with `online = 0`?

**Reply by shikokuchuo** [reply]
> Regarding online: So, if I ever see it go from 1 to 0, does that indicate something went wrong, e.g. the server terminated, or the network connection went down? Because as soon as I've got a server connected, and I don't shut it down intentionally, or it times out, it should stay at online = 1, correct?

Correct, in normal course, it should never go from 1 to 0 by itself unless it has been set to time out etc. However in case it does, i.e. it has really crashed or network has gone down, you can simply replace it with another 'instance'.

> Is it correct that:
> 
> ```r
> online <- (daemons()$daemons[, "online"] == 1L)
> ```
> 
> is relevant for knowing the _current_ state of the "cluster", whereas:
> 
> ```r
> stats <- mirai::daemons()$daemons[, c("instance", "assigned", "complete")]
> ```
> 
> is _historical_ data for what happened in the past (and irrelevant for task/worker orchestration)?
> 
> If you want to scale up and down the servers, is it correct that it's only the `online` flag you should look at, e.g. only launch servers for daemons with `online = 0`?

Absolutely - only 'online' is relevant for the current status. Only launch new server 'instances' when `online = 0`.  `crew` uses the historical data to determine when to scale up or down e.g. if some servers have hardly been used then not re-launch those when they time out.

Thank you for going over this - it is helpful to have a fresh perspective as we have been developing it intensively for so long - what is obvious to us may well not be as intuitive.

**Reply by shikokuchuo** [reply]
I found the relevant description of the status matrix in the readme - I have transferred this to the `daemons()` function documentation in 31e0413 and added some clarifications as part of 0d4e586.

I agree completely with Henrik's ideas for improving the readability of the stats, however it is currently a trade-off with efficiency as per Will's comment. Although *we* may know how to interact with data frames in a performant manner, we cannot assume this for all users, hence I prefer to maintain the matrix, which is efficient by default.
